### ISSUE TITLE
Copy configuration files

### ISSUE THREAD

**Issue by lra** [lra/mackup#613]
Linking file does not work for all applications, and some users would prefer to actually "backup" their config files, instead of linking them.

Finally, the app store applications live in a container that's not working for application linking, a copy mode would allow us to support those apps.

We should be able to specify if the mode (`link` vs `copy`) in each application INI.

As it's gonna be complex to understand what's going on if mackup supports both linking and copying, this feature depends on `mackup status` https://github.com/lra/mackup/issues/603

Finally, mackup should say goodbye to `mackup backup` and `mackup restore` and use more explicit actions, e.g. `mackup link`, `mackup copy`, ...

PR welcomed

**Comment by vitorgalvao** [comment]
For MAS apps, it’ll likely also need to `defaults read` preferences to a file and restore them with `defaults write`. This is why I’ve been mentioning storing bundle IDs.

**Comment by lra** [comment]
You are right, I think this can come as a later feature, once this one is in: https://github.com/lra/mackup/issues/615

**Comment by zanderzhng** [comment]
we may add StickiesDatabase back with this feature. #412

**Comment by martin-liu** [comment]
When I tried mackup, I thought it will copy the config files. 
So that when I try `mackup backup` and find it works, I just delete the backup folder and empty the trash.

Now all my config files LOST!!!

Why not you guys write this behavior to README? :sob:

**Comment by zanderzhng** [comment]
If your files were synced up with dropbox, you may recover them with it.

**Comment by lra** [comment]
@martin-liu it's explained: https://github.com/lra/mackup#bullsht-what-does-it-really-do-to-my-files
Feel free to propose a better explanation in a PR.
Also what @zanderzhng said, you can undelete files in dropbox.

**Comment by vitorgalvao** [comment]
@lra To be fair, it doesn’t really explain it. It says `mackup` does `ln`. If you don’t know what `ln` is (and why would you, if you just want to backup your configurations) you don’t know what’s happening. It’s also the third point in a list of technical instructions (which a user won’t read if, again, all they want to do is backup configurations).

The only time `link` is even referenced in the README is in the word “linked”, [five paragraphs into a long explanation most people won’t care about](https://github.com/lra/mackup#why-did-you-do-this) that says what you *used* to do (i.e. it doesn’t even mean that’s what `machup` does).

**Comment by martin-liu** [comment]
@zanderzhng @lra , thanks for reply. 
But my config files including ~/.emacs.d and have several hundred MB, it did not have time to sync to Dropbox. 
While luckily, I have most of the configuration in my github :smiley:

**Comment by nlemoine** [comment]
:+1: Totally agree with this topic, it would be awesome to have an copy/link choice.

**Comment by liukun** [comment]
I suggest introduce two new type of storage engine `file_system_link` `file_system_copy` to specify the behavior, and deprecate the old `file_system`.

**Comment by Vvkmnn** [comment]
I use this tool across almost 3 machines. A feature like this would be awesome.

**Comment by mattmc3** [comment]
The copy feature has been requested numerous times over the years, and as I was looking at it, it seemed an easy task to leverage the .cfg files to create a simple rsync-based copy feature with a lite bash script. Thus, Rackup was born. [https://github.com/mattmc3/rackup](https://github.com/mattmc3/rackup). Rackup leverages Mackup's configs to to an rsync based file copy backup instead of symlinking.

**Comment by vitorgalvao** [comment]
@mattmc3 Copying alone isn’t enough for Mac App Store apps. They cache their settings.

**Comment by mattmc3** [comment]
@vitorgalvao - Maybe I'm missing what you're really saying, but I'm not sure how cached settings is that big of a deal. Time machine backups only work against the files written to disk as well. When the file is written, an rsync backup can get it. If it stays cached in memory, then nothing was going to pick it up anyway. Like I said, maybe I'm missing what you're getting at, but from my vantage point, it's not a big enough issue to architect something grandiose to cover it.

**Comment by vitorgalvao** [comment]
@mattmc3 Due to the way settings are cached, if you copy the settings over in MAS (Mac App Store) apps, they aren’t read. So copying alone is ineffective. You need to *at least* do a `defaults read {{app_bundle_id}}`.

I do that in my own settings-saving script and it mostly works, but that means mackup’s `.cfg` files are not enough, as they do not contain the bundle ids.

**Comment by mattmc3** [comment]
@vitorgalvao - Good call out. Seems like we can solve this with a simple pre-backup script, which I'm going to need anyway to add other plugin scripts to settings that aren't in files like the list of Atom packages, a Brewfile, etc. To account for the issue you mention, before I make a Brewfile, I can `brew install mas`, which will add Mac App Store items to the `Brewfile`, and those contain the bundle ids. On restore, we can then loop through the Brewfile mas entries and run `defaults read {{app_bundle_id}}` from each of them. I've [opened an issue in my repo](https://github.com/mattmc3/rackup/issues/2).

**Comment by Ashtonian** [comment]
just to be clear tunnelblick is one of the applications that _needs_ this to function properly.

**Comment by bam80** [comment]
#2012

**Comment by bam80** [comment]
> Thus, Rackup was born. https://github.com/mattmc3/rackup

Broken link, what happened with it? @mattmc3

**Comment by mattmc3** [comment]
I don't know anything about a project called 'rackup', but I made something called mac.bak based off this project that just uses rsync: https://github.com/mattmc3/mac.bak

**Comment by bam80** [comment]
> I made something called mac.bak based off this project that just uses rsync: https://github.com/mattmc3/mac.bak

This seems abandoned (correct me if I'm wrong).
Meanwhile, can anyone suggest other similar tool that fulfills the requirement, maybe?

**Comment by vitorgalvao** [comment]
> Meanwhile, can anyone suggest other similar tool that fulfills the requirement, maybe?

I have since created [Tape](https://github.com/vitorgalvao/tape). It’s been serving me well for years. I’m not actively adding support for new software (just fulfilling my own needs) but I do consider pull requests.

**Comment by mattmc3** [comment]
> This seems abandoned (correct me if I'm wrong).

The project is evergreen. Update the submodule to get the lastest mackup (`make submodules`), generate the rsync files with the enclosed script (`make apps`), run the backup. I refreshed it just now, but really anyone could.

**Comment by sedinamwright330** [comment]
Ok

null

**Comment by sedinamwright330** [comment]
Ok

null

**Comment by lra** [comment]
Done in https://github.com/lra/mackup/pull/2085
